src/main/java/hudson/remoting/Launcher.java

```diff
@@ -147,16 +148,28 @@ public void setTextMode(boolean b) {
             usage="add the given classpath elements to the system classloader.")
     @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "Parameter supplied by user / administrator.")
     public void addClasspath(String pathList) throws Exception {
-        Method $addURL = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
-        $addURL.setAccessible(true);
-
-        for(String token : pathList.split(File.pathSeparator))
-            $addURL.invoke(ClassLoader.getSystemClassLoader(),new File(token).toURI().toURL());
-
-        // fix up the system.class.path to pretend that those jar files
-        // are given through CLASSPATH or something.
-        // some tools like JAX-WS RI and Hadoop relies on this.
-        System.setProperty("java.class.path",System.getProperty("java.class.path")+File.pathSeparatorChar+pathList);
+        if (ClassLoader.getSystemClassLoader() instanceof  URLClassLoader) {
+            // Java 8
+            Method $addURL = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
+            $addURL.setAccessible(true);
+
+            for(String token : pathList.split(File.pathSeparator))
+                $addURL.invoke(ClassLoader.getSystemClassLoader(),new File(token).toURI().toURL());
+
+            // fix up the system.class.path to pretend that those jar files
+            // are given through CLASSPATH or something.
+            // some tools like JAX-WS RI and Hadoop relies on this.
+            System.setProperty("java.class.path",System.getProperty("java.class.path")+File.pathSeparatorChar+pathList);
+        } else {
+            // Java 9 or later
+            List<URL> urls = new ArrayList<>();
+            for (String token : pathList.split(File.pathSeparator))
+                urls.add(new File(token).toURI().toURL());
+
+            ClassLoader currentThreadClassLoader = Thread.currentThread().getContextClassLoader();
+            URLClassLoader urlClassLoader = new URLClassLoader(urls.toArray(new URL[]{}), currentThreadClassLoader);
+            Thread.currentThread().setContextClassLoader(urlClassLoader);
```

Comment:
Do not rely on `Thread.contextClassLoader` please.

If you want to offer a flag allowing custom listener code to be loaded, do not do it this way. First of all, stop using system properties; introduce a real option with documentation. Then have it somehow specify the location of a JAR and/or class name(s) to load from it in a fresh class loader.

(Nicest would be to use SezPoz but adding a new library dep for this purpose is likely overkill, and anyway it would potentially clash with the copy loaded from `jenkins-core.jar`.)

Reply:
For example, define a hook as a JAR whose manifest has a `Main-Class` entry giving the FQN of a class implementing whichever listener interface makes sense, with a `public` no-arg constructor. Then define a CLI arg

```bash
-listener /path/to/impl.jar
```

When specified, the hook would be loaded in its own private `URLClassLoader` parented to that of `agent.jar`.